### PR TITLE
Colorimetry tool for node context

### PR DESCRIPTION
**Pull request by VibhorCodecianGupta**

### CODE REVIEW

**Comment by HarshKhandeparkar** [comment]
@VibhorCodecianGupta add grunt as a devdependency and add the default Gruntfile in the docs. I want to add the UI or at least start one. I will use grunt for it.

File: package.json [hunk 1]
```diff
@@ -0,0 +1,15 @@
+{
+  "name": "colorimetry",
+  "version": "1.0.0",
```
Comment (HarshKhandeparkar) [inline package.json]: Version can be a pre release `0.x.x`

File: package.json [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+  "main": "index.js",
+  "repository": "https://github.com/publiclab/colorimetry.git",
+  "author": "jywarren",
```
Comment (HarshKhandeparkar) [inline package.json]: Let's make this publiclab I guess.

File: package.json [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+  "license": "GNU",
```
Comment (HarshKhandeparkar) [inline package.json]: Better to say `GPL-3`

File: package.json [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+  "dependencies": {
+    "base64-img": "^1.0.4",
+    "image-sequencer": "^3.1.1"
+  },
+  "scripts": {
+    "start": "node colorimetry.js"
```
Comment (HarshKhandeparkar) [inline package.json]: Let's put everything inside a `src/` directory and add an index.js file at the start. What do you say?

**Comment by VibhorCodecianGupta** [comment]
@HarshKhandeparkar let's keep the UI as a second milestone? I think it's important we discuss the functionality first. Once that is sorted, UI will be pretty straight forward. There are some considerations here, for example, what is our use case, will we assume the input images to be perfect and well lit, will they be separate or will the test and result image be merged, is averaging the red channel a standardized colourimetry measure, etc.
As for the grunt file, I think whoever works on the UI may add it themselves, for they'll know the best flow for it.

**Comment by HarshKhandeparkar** [comment]
Ok. Np. Btw, is colorimetry only for getting difference in colors and why only red channel? 
@jywarren

**Comment by HarshKhandeparkar** [comment]
@jywarren there was another real world use case of particle analysis right? Does it include both colorimetry and finding the area? If yes then can we change this repo to a general purpose particle analysis or any real world chemical or physical analysis lib? If colorimetry only includes getting the difference between colors, creating a whole new repo for it seems kind of overkill.

**Comment by HarshKhandeparkar** [comment]
Are you working on the package file? (The changes i had suggested) or should we do those later?

**Comment by VibhorCodecianGupta** [comment]
Yes, I am. Will push some more changes in a while, working on them :)

**Comment by HarshKhandeparkar** [comment]
Ok.

**Comment by HarshKhandeparkar** [comment]
Please see gitter btw. Delete this comment once you see it.

**Comment by VibhorCodecianGupta** [comment]
@jywarren here is the foundation of the colourimetry code, in the node environment. Two sequencers are initiated, images and steps are loaded and a promise of the output is returned which is then resolved to calculate the image distance (which, for now, is the ratio of the average red channel value of the outputs of the sequencers) and to write the output images to a folder.

Terminal log:

Screen Shot 2019-04-19 at 7 43 47 PM

Note that for a browser environment, a canvas needs to be made for the sequencer to function on. Will get on with it once this is good to go.

**Comment by HarshKhandeparkar** [comment]
@jywarren can we(is-reviewers) get write(no merge) access to this repo as well?

**Comment by jywarren** [comment]
This looks great. I think we should try to use either the same "drag an
image" image select starter module as we do in the main repo, OR use the
`uppy` package which allows for importing from many sources:

(@harshithpabbati has been working on this here:
https://github.com/publiclab/image-sequencer/pull/840)

We could start with something simple like the existing
drag/drop/select/picture code, and swap in Uppy when it's ready... whatever
you like!

Eventually we'll want to show the starting image with a transparent overlay
marking the required regions so people can line it up well. Make sense?

This is looking awesome!

**Comment by VibhorCodecianGupta** [comment]
Sounds good! Shall I code the UI here itself, then?

**Comment by HarshKhandeparkar** [comment]
I would like to code the UI.

**Comment by HarshKhandeparkar** [comment]
Can I?

**Comment by VibhorCodecianGupta** [comment]
Sure, feel free to chime in! Pull the master branch of my repo and keep pushing commits there, that'll help us collaborate :D

**Comment by HarshKhandeparkar** [comment]
It won't be that difficult. Also I wanted to ask if it could use react. @VibhorCodecianGupta @jywarren
React will be simpler to use, code and read. We will also not have to think much about the dist files nor about development, linting and testing. They have it all covered. They can be deployed to gh-pages using the npm package `gh-pages`. We(me, harshith, my friend) have used it for one of our organisation website. It is a piece of cake to manage.

Links:
- [Website](https://theblazecoders.github.io)
- [Code](https://github.com/theblazecoders/theblazecoders.github.io)

**Comment by HarshKhandeparkar** [comment]
Sure, feel free to chime in! Pull the master branch of my repo and keep pushing commits there, that'll help us collaborate :D

Do you think that is needed? Can't I directly push here(open a PR). You will handle the backend so our branches can be rebased without any conflicts. I will always stay behind you(rebase my branch instead of you having to rebase yours).

**Comment by VibhorCodecianGupta** [comment]
Open a new PR if you need to, not a problem. Although, I'm not sure how @jywarren wants us to proceed. If this piece of code makes enough sense, it could be merged first and then UI could be developed, or if it doesn't, the UI can be built side by side. 
Also, for a use case like this, I don't think React would be necessary. Frontend frameworks have a lot of overhead which might be undesirable if our use case is not very heavy. For now, I really think a couple of functionalities like uploading/dragging images and displaying the results is all we need.
cc @jywarren

**Comment by HarshKhandeparkar** [comment]
I am not going to code until Sunday so lets wait to hear what @jywarren says. You can work on the initial algorithm. After that is merged, I will work on the UI.

**Comment by HarshKhandeparkar** [comment]
> Also, for a use case like this, I don't think React would be necessary. Frontend frameworks have a lot of overhead which might be undesirable if our use case is not very heavy. For now, I really think a couple of functionalities like uploading/dragging images and displaying the results is all we need.

They do but they also compile only the required files and also minify them. I still don't know if this repo is going to be this simple. Can we add extra stuff to this repo?

> @jywarren there was another real world use case of particle analysis right? Does it include both colorimetry and finding the area? If yes then can we change this repo to a general purpose particle analysis or any real world chemical or physical analysis lib? If colorimetry only includes getting the difference between colors, creating a whole new repo for it seems kind of overkill.

Can we generalise the repo instead of specialising it to this extent? @jywarren . I don't know if I am fully correct.

**Comment by jywarren** [comment]
Shall we merge this PR first, then develop UI in another?

Re: React, I think it could be a nice future step, but I agree with @VibhorCodecianGupta that we should start with a really minimal implementation that is as small an amount of code as possible, then weigh the pros and cons. This could just be a file select and a number in the first version. But making it nicer /after/ that initial step would be great!

**Comment by HarshKhandeparkar** [comment]
Ok. Merge this. But what about

> @jywarren there was another real world use case of particle analysis right? Does it include both colorimetry and finding the area? If yes then can we change this repo to a general purpose particle analysis or any real world chemical or physical analysis lib? If colorimetry only includes getting the difference between colors, creating a whole new repo for it seems kind of overkill.

**Comment by jywarren** [comment]
> @jywarren there was another real world use case of particle analysis right? Does it include both colorimetry and finding the area? If yes then can we change this repo to a general purpose particle analysis or any real world chemical or physical analysis lib? If colorimetry only includes getting the difference between colors, creating a whole new repo for it seems kind of overkill.

There is another use case which is particle analysis, and there's a long issue and discussion on this in image-sequencer. I think it may need its own approach, and could be prototyped within image-sequencer. 

The reasons for a separate repo on this are:

1. it required more than one sequencer (even so, i think we could've made it in the /examples/ folder as `/examples/colorimetry.html`, but could go either way)
2. we can host the demo at a very simple GitHub Pages URL from here once done (which'd be harder from the /examples/ folder)
3. we want a very smooth, easy interface for people to use this -- imagine using it from your phone, what would make it super easy and intuitive and simple? 
4. isolating it from image-sequencer stabilizes it because we source via NPM instead of the leading edge commit

I hope this makes sense! Thanks, all!

**Comment by HarshKhandeparkar** [comment]
Yes it does make sense!

**Comment by HarshKhandeparkar** [comment]
> @jywarren can we(is-reviewers) get write(no merge) access to this repo as well?

**Comment by VibhorCodecianGupta** [comment]
@jywarren sounds good. Let's get this merged. I'll look for some standard ways to quantify colourimetry, I was reading up some research papers earlier today for mathematical algorithms. @HarshKhandeparkar and I can create the barebones UI, and a long term project could be the react implementation. Cheers!

**Comment by HarshKhandeparkar** [comment]
I think completely aceraging colors isn't a good idea. What do you say? We have to add more weight to the more/most dominant color. Improving the algorithm can be our next goal.

**Comment by HarshKhandeparkar** [comment]
@jywarren we have merge rights? Didnt you add branch protection rules? Or is this done on purpose?

**Comment by VibhorCodecianGupta** [comment]
Now using the new input support, namely `crop{x:130|y:115|w:50|h:50},average{}`

**Comment by jywarren** [comment]
yeah it's great! Vibhor, is this ready to merge then? Thanks!

**Comment by HarshKhandeparkar** [comment]
@jywarren I actually could edit the settings and dependabot added new commits. This might be insecure. Please have a look at it. I don't think I should have been able to add dependabot to this repo.

**Pull request merged**